qubes-dist-upgrade.sh

```diff
@@ -537,96 +192,30 @@ max_concurrency="${max_concurrency:-4}"
-        if ! [ -e "/var/lib/qubes/vm-kernels/$new_kernel" ]; then
-            echo "ERROR: Kernel $new_kernel installed but /var/lib/qubes/vm-kernels/$new_kernel is missing!"
-            exit 1
-        fi
-        echo "Changing default kernel from $default_kernel to $new_kernel"
-        qubes-prefs default-kernel "$new_kernel"
-    fi
-
-    exit 0
-fi
-
-trap 'exit_migration' 0 1 2 3 6 15
 # shellcheck disable=SC1003
-echo 'WARNING: /!\ ENSURE TO HAVE MADE A BACKUP OF ALL YOUR VMs AND dom0 DATA /!\'
+echo 'WARNING: /!\ MAKE SURE YOU HAVE MADE A BACKUP OF ALL YOUR VMs AND dom0 DATA /!\'
 if [ "$assumeyes" == "1" ] || confirm "-> Launch upgrade process?"; then
-    # Backup xen and kernel cmdline
-    cat /proc/cmdline > /tmp/kernel_cmdline
-    if ! [ -f /tmp/xen_cmdline ]; then
-        xl info xen_commandline > /tmp/xen_cmdline || {
-            echo "ERROR: Failed to get Xen cmdline, have you restarted the system after previous upgrade stage?"
-            exit 1
-        }
-        if ! grep -q gnttab_max_frames /tmp/xen_cmdline; then
-            sed -e 's:$: gnttab_max_frames=2048 gnttab_max_maptrack_frames=4096:' -i /tmp/xen_cmdline
-        fi
-    fi
-
     # Shutdown nonessential VMs
     shutdown_nonessential_vms
 
-    pool_name=$(get_thin_pool_name)
-    if [ "$double_metadata_size" == 1 ]; then
-        if [ -z "$pool_name" ]; then
-            echo "---> (STAGE 0) Skipping - no LVM thin pool found"
-        elif [ "$(get_tmeta_size "$pool_name")" -ge "$(recommended_size "$pool_name")" ]; then
-            echo "---> (STAGE 0) Skipping - already right size"
-        else
-            echo "---> (STAGE 0) Adjusting LVM thin pool metadata size..."
-            set_tmeta_size "$pool_name"
-        fi
-    fi
-
-    if [ "$update" = "1" ] && [ "$(rpm -q --qf='%{VERSION}' qubes-release)" = "4.1" ]; then
+    if [ "$update" = "1" ] && [ "$(rpm -q --qf='%{VERSION}' qubes-release)" = "4.2" ]; then
         echo "---> (STAGE 1) Updating dom0... already done, skipping"
         update=
     fi
 
     if [ "$update" == "1" ]; then
+      root_vol_name=$(get_root_volume_name)
+      root_group_name=$(get_root_group_name)
+      if [ -z "$root_vol_name" ]; then
+        echo "---> (STAGE 1) Skipping dom0 snapshot - no LVM volume found"
+      elif lvs "$root_group_name/Qubes41UpgradeBackup" > /dev/null 2>&1 ; then
+        echo "---> (STAGE 1) Skipping dom0 snapshot - snapshot already exists. If you want to make a snapshot anyway, remove the existing one using lvremove $root_group_name/Qubes41UpgradeBackup"
+      elif [ "$assumeyes" == "1" ] || confirm "---> (STAGE 1) Do you want to make a dom0 snapshot?"; then
+        # make a dom0 snapshot
+        lvcreate -n Qubes41UpgradeBackup -s "$root_vol_name"
+        echo "--> If upgrade to 4.2 fails, you can restore your dom0 snapshot after booting from a bootable device and using sudo lvconvert --merge $root_group_name/Qubes41UpgradeBackup. Reboot after restoration."
```

Comment:
It seems I was wrong, it is possible to do that from running system, and then reboot.